### PR TITLE
Introducing the use of macros for DPIA primitives

### PR DESCRIPTION
**Pull request by michel-steuwer**
This PR introduces three new macros `@expPrimitive`, `@accPrimitive`, and `@cmdPrimitive` that generate boilerplate code for DPIA primitives (including the crucial `visitAndRebuild` method).

Overall this PR removes over 2000 lines of code :-)

### CODE REVIEW

File: src/main/scala/shine/DPIA/primitives/intermediate/SlideSeqIValues.scala [hunk 1]
```diff
@@ -38,9 +38,9 @@ object SlideSeqIValues {
...
           k(rs.rd) `;`
           // rotate
-          MapSeqI(size - 1, dt1, dt2, load,
+          MapSeqI(unroll = true)(size - 1, dt1, dt2, load,
             Drop(1, size - 1, dt1, rs.rd),
-            TakeAcc(size - 1, 1, dt2, rs.wr), unroll = true)
+            TakeAcc(size - 1, 1, dt2, rs.wr))
         ),
       arithexpr.arithmetic.RangeAdd(0, n, 1)))
     }))
```
Comment (Bastacyclop) [inline src/main/scala/shine/DPIA/primitives/intermediate/SlideSeqIValues.scala:46]: maybe this should become `RotateValuesI` for consistency.

File: src/main/scala/shine/DPIA/primitives/functional/CircularBuffer.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  def streamTranslation(C: Phrase[`(nat)->:`[(ExpType ->: CommType) ->: CommType] ->: CommType])
+                       (implicit context: TranslationContext): Phrase[CommType] = {
+    val i = NatIdentifier(freshName("i"))
+    str(input)(fun((i: NatIdentifier) ->:
+      (expT(dt1, read) ->: (comm: CommType)) ->: (comm: CommType)
+    )(nextIn =>
+      SlideSeqIIndices(n, sz, 1, dt1, dt2,
+        fun(expT(dt1, read))(x =>
+          fun(accT(dt2))(o => acc(load(x))(o))),
+        nextIn, C)
```
Comment (Bastacyclop) [inline src/main/scala/shine/DPIA/primitives/functional/CircularBuffer.scala:35]: maybe this should become `CircularBufferI` for consistency.

File: src/main/scala/shine/OpenCL/CodeGenerator.scala [hunk 3]
```diff
@@ -183,10 +184,10 @@ class CodeGenerator(override val decls: CCodeGenerator.Declarations,
 
     case IdxVec(_, _, i, e) => CCodeGen.codeGenIdx(i, e, env, path, cont)
 
-    case OpenCLFunction(name, _, _, args) =>
+    case OpenCLFunctionCall(name, _, _, args) =>
```
Comment (Bastacyclop) [inline src/main/scala/shine/OpenCL/CodeGenerator.scala:187]: should this be `ocl.FunctionCall`?

File: src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala [hunk 4]
```diff
@@ -60,34 +62,38 @@ object InsertMemoryBarriers {
...
+              visitLoopBody(body, allocs, metadata, outer_wg_writes))), pf.init, pf.step))
+        case pf@ParFor(level, dim, unroll) =>
+          val (x, o, body) = pf.unwrapBody
+          Stop(ParFor(level, dim, unroll)(pf.n, pf.dt, pf.out,
+            Lambda(x, Lambda(o, visitLoopBody(body, allocs, metadata))), pf.init, pf.step))
+        case pf@ParForNat(level, dim, unroll) =>
+          val (x, o, body) = pf.unwrapBody
+          Stop(ParForNat(level, dim, unroll)(pf.n, pf.ft, pf.out,
+            DepLambda[NatKind, AccType ->: CommType](x,
+              Lambda(o, visitLoopBody(body, allocs, metadata))), pf.init, pf.step))
```
Comment (Bastacyclop) [inline src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala:87]: looks like the previous `ParForNat(Local` case is missing.

File: src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala [hunk 5]
```diff
@@ -60,34 +62,38 @@ object InsertMemoryBarriers {
...
+        case OclNew(addr, _, Lambda(x, _)) if addr != AddressSpace.Private =>
           Continue(p, Visitor(allocs + (x -> addr), metadata))
-        case OpenCLNewDoubleBuffer(addr, dt1, dt2, dt3, n, in, out, Lambda(x, body))
+        case OclNewDoubleBuffer(addr, dt1, dt2, dt3, n, in, out, Lambda(x, body))
```
Comment (Bastacyclop) [inline src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala:90]: `ocl.NewDoubleBuffer` for consistency?

**Review by Bastacyclop — COMMENTED** [review]
Looks great! I only have minor comments, also I think that the line width goes over 100 in multiple spots.

**Comment by Bastacyclop** [comment]
oh, and should we use this opportunity to rename the Shine packages? maybe `shine.dpia`, `shine.opencl`, `shine.openmp`, .. like we did for Rise.

File: macros/src/main/scala/shine/macros/Primitive.scala [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+    def expPrimitive(annottees: c.Expr[Any]*): c.Expr[Any] = {
+      annottees.map(_.tree) match {
+        case (cdef: ClassDef) :: Nil =>
+          c.Expr(expPrimitivesFromClassDef(cdef))
+        case (cdef: ClassDef) :: (md: ModuleDef) :: Nil =>
+          c.Expr(q"{${expPrimitivesFromClassDef(cdef)}; $md}")
+        case _ => c.abort(c.enclosingPosition, "expected a class definition")
+      }
+    }
```
Comment (umazalakain) [inline macros/src/main/scala/shine/macros/Primitive.scala:35]: Could these three `*Primitive` functions be made into one function that accepts the `ClassDef => ClassDef` transformations as a a parameter?

File: macros/src/main/scala/shine/macros/Primitive.scala [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+          body.asInstanceOf[List[Tree]],
+          parents.asInstanceOf[List[Tree]])
+      case _ =>
+        c.abort(c.enclosingPosition, "expected a case class extends Primitive")
+    }
+
+    def makeAccPrimitiveClass(name: TypeName,
+                              additionalParams: List[ValDef],
+                              params: List[ValDef],
+                              body: List[Tree]): ClassDef = {
```
Comment (umazalakain) [inline macros/src/main/scala/shine/macros/Primitive.scala:254]: Why does `makeExpPrimitiveClass` handle parents but this and `makeComPrimitiveClass` don't?

File: macros/src/main/scala/shine/macros/Primitive.scala [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+         """
+      }).asInstanceOf[ClassDef]
+
+      c.info(c.enclosingPosition,
+        s"generated `${name.toString}'\n$expClass", force = false)
+
+      expClass
+    }
+
+    def expPrimitivesFromClassDef: ClassDef => ClassDef = {
```
Comment (umazalakain) [inline macros/src/main/scala/shine/macros/Primitive.scala:216]: Would it be possible to have a single `*PrimitivesFromClassDef` that accepts the `make*PrimitiveClass` as an argument?

File: macros/src/main/scala/shine/macros/Primitive.scala [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+        makeVisitAndRebuild(name, additionalParams, params)
+      else q""}
+
+          ${if (xmlPrinterMissing)
+        makeXMLPrinter(name, additionalParams, params)
+      else q""}
+         """
+
+      val accClass = (additionalParams match {
+        case List() =>
```
Comment (umazalakain) [inline macros/src/main/scala/shine/macros/Primitive.scala:276]: This class generation bit seems common to all three cases, modulo the extended class and handling parents

File: macros/src/main/scala/shine/macros/Primitive.scala [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+      }
+    }
+
+    def makeLowerCaseName(s: String): String =
+      s"${Character.toLowerCase(s.charAt(0))}${s.substring(1)}"
+
+    def makeVisitAndRebuild(name: TypeName,
+                            additionalParams: List[ValDef],
+                            params: List[ValDef]): Tree = {
+      val v = q"v"
```
Comment (umazalakain) [inline macros/src/main/scala/shine/macros/Primitive.scala:63]: Is this to highlight that the generated code works with name `v`?

File: macros/src/main/scala/shine/macros/Primitive.scala [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+            q"shine.DPIA.Phrases.VisitAndRebuild($name, $v)"
+          // Vector[Phrase[ExpType]]
+          case AppliedTypeTree((Ident(TypeName("Vector")),
+          List(AppliedTypeTree((Ident(TypeName("Phrase")), _)))))
+               |   AppliedTypeTree((Ident(TypeName("Seq")),
+          List(AppliedTypeTree((Ident(TypeName("Phrase")), _)))))
+          =>
+            q"$name.map(shine.DPIA.Phrases.VisitAndRebuild(_, $v))"
+          case _ =>
+            q"$name"
```
Comment (umazalakain) [inline macros/src/main/scala/shine/macros/Primitive.scala:93]: Could this be reified into normal code? Could we have a function that matches on the kind and returns the appropriate visitor function?

**Review by umazalakain — COMMENTED** [review]
Quite an impressive feat!

**Comment by umazalakain** [comment]
Removes 2500 LoC now ;)

File: src/main/scala/rise/core/primitives.scala [hunk 12]
```diff
@@ -260,15 +260,15 @@ object primitives {
     impl{ n: Nat => expl((alloc: Nat) => expl((sz: Nat) =>
       impl{ s: DataType => impl{ t: DataType =>
         (s ->: t) ->: // function to load an input
-          ((n + sz) `.` s) ->: ((1 + n) `.` sz `.` t) }}))}
+          ((n - 1 + sz) `.` s) ->: (n `.` sz `.` t) }}))}
```
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply src/main/scala/rise/core/primitives.scala:263]: This brings it in line with the DPIA definition of the primitive (before the meaning of `n` changes between rise and dpia)

File: src/main/scala/rise/core/primitives.scala [hunk 13]
```diff
@@ -260,15 +260,15 @@ object primitives {
...
   }
 
   // mainly to achieve register rotation
   @primitive object rotateValues extends Primitive with Builder {
     // TODO: should return a stream / sequential array, not an array
     impl{ n: Nat => expl((sz: Nat) => impl{ s: DataType =>
       (s ->: s) ->: // function to write a value
-        ((n + sz) `.` s) ->: ((1 + n) `.` sz `.` s) })}
+        ((n - 1 + sz) `.` s) ->: (n `.` sz `.` s) })}
```
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply src/main/scala/rise/core/primitives.scala:271]: same :-)

File: src/main/scala/shine/DPIA/primitives/functional/IndexAsNat.scala [hunk 14]
```diff
@@ -1,51 +1,23 @@
...
-
-  def visitAndRebuild(fun: VisitAndRebuild.Visitor): Phrase[ExpType] =
-    IndexAsNat(fun.nat(n), VisitAndRebuild(e, fun))
-
-  def eval(s: OperationalSemantics.Store): OperationalSemantics.Data = {
-    OperationalSemantics.eval(s, e) match {
-      case IndexData(i, _) => NatData(i)
-      case d => throw new Exception(s"Expected IndexData but found $d.")
-    }
-  }
```
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/primitives/functional/IndexAsNat.scala:36]: I think that an oversight. @umazalakain could you safe this eval implementations please.

File: src/main/scala/shine/DPIA/primitives/intermediate/SlideSeqIValues.scala (hunk 1 above)
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/primitives/intermediate/SlideSeqIValues.scala:46]: You mean renaming the file? Yes. We could do that.

File: src/main/scala/shine/DPIA/primitives/functional/CircularBuffer.scala (hunk 2 above)
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/primitives/functional/CircularBuffer.scala:35]: Same here

File: src/main/scala/shine/OpenCL/CodeGenerator.scala (hunk 3 above)
Comment (michel-steuwer) [reply src/main/scala/shine/OpenCL/CodeGenerator.scala:187]: It is a call to built in OpenCL functions (so it's not a "function call in OpenCL" but rather a call to an "OpenCL function").
That's why I kept the name as it should not just be reflected by the package its in.

Does that make sense?

File: src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala (hunk 4 above)
Comment (michel-steuwer) [reply src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala:87]: This wasn't implemented before (just a `???`) but this case should be analogous to the `ParFor(Local` case. Which indeed seems missing. I'll add it.

File: src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala (hunk 5 above)
Comment (michel-steuwer) [reply src/main/scala/shine/OpenCL/compilation/InsertMemoryBarriers.scala:90]: Fine with me. I think this is just an import alias in this file.

**Comment by michel-steuwer** [comment]
> Removes 2500 LoC now ;)

Nice :-)

File: src/main/scala/shine/DPIA/primitives/functional/IndexAsNat.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/primitives/functional/IndexAsNat.scala:36]: Actually, I'll add it back myself.

File: src/main/scala/shine/OpenCL/CodeGenerator.scala (hunk 3 above)
Comment (Bastacyclop) [reply src/main/scala/shine/OpenCL/CodeGenerator.scala:187]: I see, I think that I have personally always been using regular C function call primitives even for OpenCL built-in functions such as `dot`.

**Review by Bastacyclop — APPROVED** [review]

**Pull request merged**